cdk/src/open-data-platform/data-plane/schema.ts

```diff
@@ -0,0 +1,125 @@
+// [1] https://docs.aws.amazon.com/cdk/api/v1/docs/@aws-cdk_aws-lambda-nodejs.NodejsFunction.html#entry
+const handlerId = 'handler';
+
+// This construct works by defining a lambda that connects to the DB and executes the provided SQL.
+export class Schema extends Construct {
+  constructor(scope: Construct, id: string, props: SchemaProps) {
+    super(scope, id);
+
+    const { cluster, vpc, db, schemaFileName, credentialsSecret, userCredentials } = props;
+
+    const initSchemaFunction = new NodejsFunction(this, handlerId, {
+      vpc: vpc,
+      vpcSubnets: {
+        subnetType: ec2.SubnetType.PRIVATE_WITH_NAT,
+      },
+      environment: {
+        CREDENTIALS_SECRET: credentialsSecret.secretArn,
+        DATABASE_NAME: db,
+      },
+      timeout: Duration.minutes(5),
+      // We may need to install a dependency at the root dir of this repo to make bundling work.
+      // https://docs.aws.amazon.com/cdk/api/v1/docs/aws-lambda-nodejs-readme.html#local-bundling
+      bundling: {
+        nodeModules: ['@databases/pg'],
+        commandHooks: {
+          beforeBundling(inputDir: string, outputDir: string): string[] {
+            return [
+              `echo "Copying ${inputDir}/${schemaFileName} to ${outputDir}/schema.sql"`,
+              `cp ${inputDir}/${path.relative(
+                process.cwd(),
+                __dirname,
+              )}/${schemaFileName} ${outputDir}/schema.sql`,
+            ];
+          },
+          afterBundling() {
+            return [];
+          },
+          beforeInstall() {
+            return [];
+          },
+        },
+      },
+    });
+
+    // Set up access control for the lambda.
+    credentialsSecret.grantRead(initSchemaFunction);
+    userCredentials?.forEach((secret) => secret.grantRead(initSchemaFunction));
+    cluster.connections.allowFrom(initSchemaFunction, ec2.Port.tcp(cluster.clusterEndpoint.port));
+
+    // Invoke the lambda.
+    const init = new ResourceInitializer(this, 'InitSchema', {
+      initFunction: initSchemaFunction,
+      payload: {
+        userCredentials: userCredentials?.map((secret) => secret.secretArn),
+      },
+    });
+    init.node.addDependency(cluster);
+  }
+}
+
+// Construct that invokes a lambda one time during `cdk deploy`.
+// If there are other lambdas that we want to invoke during deployment, consider breaking this
+// out into a common file.
+//
+// Based on: https://github.com/BlueConduit/tributary/blob/main/cdk/lib/resource-initializer.ts
+interface ResourceInitializerProps {
+  initFunction: lambda.Function;
+  payload?: object;
+}
+
+class ResourceInitializer extends Construct {
+  constructor(scope: Construct, id: string, props: ResourceInitializerProps) {
+    super(scope, id);
+
+    const { initFunction, payload } = props;
+
+    const defaultPayload = {
+      source: 'customresource.init',
+    };
+
+    const apiCall: cr.AwsSdkCall = {
```

Comment:
What exactly is this doing?

Reply:
This defines a custom resource that tells CDK to invoke a lambda during deployment. I've added a few more comments to describe the code.